lib/runtime/formatter.js

```diff
@@ -66,17 +68,41 @@ module.exports = class Formatter extends FormatUtils {
     _replaceInString(str, argMap) {
         if (typeof str !== 'string')
             return undefined;
+        if (this._missingAllParams(str, argMap))
+            return null;
         return str.replace(/\$(?:\$|([a-zA-Z0-9_]+(?![a-zA-Z0-9_]))|{([a-zA-Z0-9_]+)(?::(%|[a-zA-Z-]+))?})/g, (match, param1, param2, opt) => {
             if (match === '$$')
                 return '$';
             let param = param1 || param2;
-            if (param.startsWith('{'))
-                param = param.substring(0, param.length-1);
             let value = argMap[param];
             return this._displayValue(value, opt);
         });
     }
 
+    _missingAllParams(str, argMap) {
+        const reg = /\$(?:([a-zA-Z0-9_]+(?![a-zA-Z0-9_]))|{([a-zA-Z0-9_]+)(?::(%|[a-zA-Z-]+))?})/g;
+        let match = reg.exec(str);
+        if (match === null)
+            return false; // if there is no param at all, there is no param missing
+        let existSome = false;
+        while (match !== null) {
+            const param = match[1] || match[2];
+            if (!this._isNull(argMap[param]))
+                existSome = true;
+            match = reg.exec(str);
+        }
+        return !existSome;
+    }
+
+    _isNull(value) {
+        if (value === undefined || value === null || value === [])
```

Comment:
`value === []` is buggy, as you're comparing objects by address so it will always be false. Maybe `Array.isArray(value) && value.length === 0` ?